### PR TITLE
Google validation

### PR DESCRIPTION
**Pull request by lhamu**
**Google Sign-in validation**

- Add the verification middleware that verifies the authenticity of the user with the Google Server using the Google tokenId received from the front-end.

- Usethe google-auth-library to authenticate the user.

- Store the received profile information in the users table.

### CODE REVIEW

File: src/middlewares/verifyGoogleToken.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+let auth = new GoogleAuth();
+
+/**
+ * Verify the user with google server using google-auth-library.
+ * 
+ * @param {Object} req 
+ * @param {Object} res 
+ * @param {function} next 
+ */
+
```
Comment (deepsadhi) [inline src/middlewares/verifyGoogleToken.js]: Remove blank line

File: src/models/Service.js [hunk 2]
```diff
@@ -56,16 +56,18 @@ class Service extends db.Model {
...
       logger().error('Error while persisting the service into database', err);
     }
-    if (ifMatch != null) {
+    if (ifMatch !== null) {
       logger().info('Fetching the services of projects of project id', {
         projectId
       });
       try {
         let results = await Service.where({ project_id: projectId }).fetchAll();
         let data = [];
```
Comment (deepsadhi) [inline src/models/Service.js:66]:
```js
let data = results.map(result => result.toJSON())
```
I guess this should also work
```js
let data = results.toJSON();
```

File: src/middlewares/verifyGoogleToken.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import GoogleAuth from 'google-auth-library';
+import HttpStatus from 'http-status-codes';
```
Comment (deepsadhi) [inline src/middlewares/verifyGoogleToken.js]: Order

File: src/middlewares/verifyGoogleToken.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  client.verifyIdToken(token, clientId, (e, login) => {
+    payload = login.getPayload();
+    userId = payload['sub'];
+  });
+
+  if (payload) {
+    data = {
+      id: userId,
+      name: payload['name'],
```
Comment (deepsadhi) [inline src/middlewares/verifyGoogleToken.js]:
```js
name: payload.name,
email: payload.email,
image: payload.picture
```

File: src/models/Service.js [hunk 5]
```diff
@@ -56,16 +56,18 @@ class Service extends db.Model {
     } catch (err) {
       logger().error('Error while persisting the service into database', err);
     }
-    if (ifMatch != null) {
+    if (ifMatch !== null) {
```
Comment (deepsadhi) [inline src/models/Service.js]: Add one line above

File: src/models/Service.js [hunk 6]
```diff
@@ -56,16 +56,18 @@ class Service extends db.Model {
...
       logger().info('Fetching the services of projects of project id', {
         projectId
       });
       try {
```
Comment (deepsadhi) [inline src/models/Service.js]: Add one line above

File: src/models/Service.js [hunk 7]
```diff
@@ -84,13 +86,14 @@ class Service extends db.Model {
     } catch (err) {
       logger().error('Error while persisting the service into database', err);
     }
-    if (ifMatch != null) {
+    if (ifMatch !== null) {
```
Comment (deepsadhi) [inline src/models/Service.js]: Leave one line between block of codes

File: src/scripts/project.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+/**
+ * Update a project from projectId.
+ *
+ * @param  {string|Number}  id
+ * @return {Promise}
+ */
+export async function updateProject(userId, projectId, data) {
+  // return Project.updateProject(userId,projectId,req.body);
```
Comment (deepsadhi) [inline src/scripts/project.js]: Remove unused comments

**Review by deepsadhi — CHANGES_REQUESTED** [review]
Leave blank line between statements block

File: package.json [hunk 9]
```diff
@@ -46,6 +47,7 @@
     "passport": "^0.4.0",
     "passport-google-oauth": "^1.0.0",
     "pg": "^7.1.2",
+    "regex": "^0.1.1",
```
Comment (mesaugat) [inline package.json]: Where do we need this?

File: src/migrations/20170516034519_create_default_config_table.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+exports.up = function(knex) {
```
Comment (mesaugat) [inline src/migrations/20170516034519_create_default_config_table.js]: `export function up`

File: src/migrations/20170516034519_create_default_config_table.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  return knex.schema.createTable('default_config', table => {
+    table.increments('id').primary();
+    table.string('type').notNullable();
+    table.jsonb('config').notNullable();
+    table.timestamp('created_at');
+    table.timestamp('updated_at');
+  });
+};
+
+exports.down = function(knex) {
```
Comment (mesaugat) [inline src/migrations/20170516034519_create_default_config_table.js]: `export function down`

File: src/schema/project.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import Joi from 'joi';
+
+export default {
+  name: Joi.string()
```
Comment (mesaugat) [inline src/schema/project.js]: Put in on the same line. No need to separate if the whole line doesn't cross 80 characters. 

`Joi.string().alphanum().required()`

File: src/schema/services.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import Joi from 'joi';
+
+const urlRegExp = /^((https?):\/\/)?(www.)?[a-z0-9]+(\.[a-z]{2,}){1,3}(#?\/?[a-zA-Z0-9#]+)*\/?(\?[a-zA-Z0-9-_]+=[a-zA-Z0-9-%]+&?)?$/;
+
+export default {
+  name: Joi.string()
```
Comment (mesaugat) [inline src/schema/services.js]: Same here

File: src/models/Service.js [hunk 14]
```diff
@@ -56,16 +56,19 @@ class Service extends db.Model {
     } catch (err) {
       logger().error('Error while persisting the service into database', err);
     }
-    if (ifMatch != null) {
+
+    if (ifMatch !== null) {
```
Comment (mesaugat) [inline src/models/Service.js]: The name `ifMatch` is not that good. We should change it.

File: src/validators/project.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
+import { validate } from '../utils/validator';
+import projectSchema from '../schema/project';
+
+/**
+ * Validate the post data to add a new Project
+ *
+ * @param req
```
Comment (mesaugat) [inline src/validators/project.js]: Type of param is missing.

```js
/**
 * @param {Object} req
 */
```

File: src/validators/service.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { validate } from '../utils/validator';
+import serviceSchema from '../schema/services';
+
+/**
+ * Validate the post request data of a new Service
+ *
+ * @param req
+ * @param res
```
Comment (mesaugat) [inline src/validators/service.js]: Here too.

**Review by mesaugat — CHANGES_REQUESTED** [review]

**Pull request merged**